### PR TITLE
chore(v2): generate cache key and hash key

### LINKED ISSUE #5816: [v2compat] generate cache key and calculate hash of the cache key
### What feature would you like to see?

generate cache key and calculate hash of the cache key.
key. 
Specifically the cache key include:
- the mapping between the input artifact names and MLMD artifact ids
- the mapping between input parameter name and input parameter value 
- the mapping between output artifact name and output artifact spec
- the mapping between output parameter name and output parameter spec
- the full execution template/spec.

We will follow [KFP V1](https://github.com/kubeflow/pipelines/blob/43994abb7e07c71084d7b0fbebab8c9e6b7aa33f/backend/src/cache/server/mutation.go#L252) which uses the [sha256](https://pkg.go.dev/crypto/sha256) to generate the hash of the cache key.
### Is there a workaround currently?

---

Love this idea? Give it a 👍. We prioritize fulfilling features with the most 👍.

### PR DESCRIPTION
**Pull request by capri-xiyue**
**Description of your changes:**
Fixed #5816, generate cache key and hash key

**Checklist:**
- [ ] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

File: v2/component/launcher.go [hunk 1]
```diff
@@ -482,6 +495,78 @@ func localPathForURI(uri string) (string, error) {
...
+		image:    l.options.ContainerImage,
+		cmdArgs : cmdAgrs,
+	}
+
+	return &cacheKey, nil
+
+}
+
+func (l *Launcher) generateFingerPrint(cacheKey cacheKey) (string, error) {
+	b, err := json.Marshal(cacheKey)
```
Comment (Bobgy) [inline v2/component/launcher.go:557]: Can this guarantee unordered items have a stable order? e.g. different input parameter names

**Review by Bobgy — COMMENTED** [review]
Shall we move cache related code to a new cache folder? We will want to reuse them, and launcher already includes too much.

I feel a little reluctant for skipping unit testing. E.g. the unordered item issue is subtle and may be hard to catch.

**Review by Bobgy — COMMENTED** [review]
FYI, you can use this executor input as reference https://github.com/kubeflow/pipelines/blob/039e8a6fe601fcc0e3d23815d476e2d68e1de081/v2/component/runtime_info_test.go#L234

File: v2/component/launcher.go (hunk 1 above)
Comment (capri-xiyue) [reply v2/component/launcher.go:557]: The json package always orders keys when marshalling. Specifically:

1. Maps have their keys sorted lexicographically
2. Structs keys are marshalled in the order defined in the struct

FYI: https://stackoverflow.com/questions/18668652/how-to-produce-json-with-sorted-keys-in-go and https://golang.org/pkg/encoding/json/#Marshal.

**Comment by capri-xiyue** [comment]
> Shall we move cache related code to a new cache folder? We will want to reuse them, and launcher already includes too much.
> 
> I feel a little reluctant for skipping unit testing. E.g. the unordered item issue is subtle and may be hard to catch.

The json marshall package handles the unordered item issue. I don't think we need UT for now to verify the functionality of json marshall package. In addition, same logic are used in KFP V1 https://github.com/kubeflow/pipelines/blob/a44a22559b511e57e6471339a264795f6d80b639/backend/src/cache/server/mutation.go#L223.

File: v2/component/launcher.go [hunk 2]
```diff
@@ -482,6 +493,64 @@ func localPathForURI(uri string) (string, error) {
 	return "", fmt.Errorf("found URI with unsupported storage scheme: %s", uri)
 }
 
+func (l *Launcher) generateCacheKey(
+	inputs *pipeline_spec.ExecutorInput_Inputs,
+	outputs *pipeline_spec.ExecutorInput_Outputs,
+	cmdAgrs []string) (*cache_utils.CacheKey, error) {
```
Comment (Bobgy) [inline v2/component/launcher.go:499]: nit: typo

File: v2/cache_utils/cache.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
+package cache_utils
```
Comment (Bobgy) [inline v2/cache_utils/cache.go:1]: nit: go styleguide is to avoid using `_`

**Review by Bobgy — COMMENTED** [review]
/lgtm

File: v2/component/launcher.go (hunk 2 above)
Comment (capri-xiyue) [reply v2/component/launcher.go:499]: done

File: v2/cache_utils/cache.go (hunk 3 above)
Comment (capri-xiyue) [reply v2/cache_utils/cache.go:1]: done

**Comment by Bobgy** [comment]
/lgtm
/approve
Looks great!

**Pull request merged**